presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,424 @@
+            if (bestResult == null) {
+                checkState(sources.size() > 1, "sources size is less than or equal to one");
+                ImmutableList.Builder<JoinEnumerationResult> resultBuilder = ImmutableList.builder();
+                Set<Set<Integer>> partitions = generatePartitions(sources.size()).collect(toImmutableSet());
+                for (Set<Integer> partition : partitions) {
+                    JoinEnumerationResult result = createJoinAccordingToPartitioning(sources, outputSymbols, partition);
+                    if (result.cost.hasUnknownComponents()) {
+                        memo.put(multiJoinKey, result);
+                        return result;
+                    }
+                    if (!result.cost.equals(INFINITE_COST)) {
+                        resultBuilder.add(result);
+                    }
+                }
+
+                List<JoinEnumerationResult> results = resultBuilder.build();
+                if (results.isEmpty()) {
+                    memo.put(multiJoinKey, INFINITE_COST_RESULT);
+                    return INFINITE_COST_RESULT;
+                }
+
+                bestResult = resultComparator.min(resultBuilder.build());
+                memo.put(multiJoinKey, bestResult);
+            }
+            if (bestResult.planNode.isPresent()) {
+                log.debug("Least cost join was: " + bestResult.planNode.get().toString());
+            }
+            return bestResult;
+        }
+
+        /**
+         * This method generates all the ways of dividing totalNodes into two sets
+         * each containing at least one node. It will generate one set for each
+         * possible partitioning. The other partition is implied in the absent values.
+         * In order not to generate the inverse of any set, we always include the 0th
+         * node in our sets.
+         *
+         * @return A set of sets each of which defines a partitioning of totalNodes
+         */
+        @VisibleForTesting
+        static Stream<Set<Integer>> generatePartitions(int totalNodes)
+        {
+            checkArgument(totalNodes >= 2, "totalNodes must be greater than or equal to 2");
+            Set<Integer> numbers = IntStream.range(0, totalNodes)
+                    .boxed()
+                    .collect(toImmutableSet());
+            return Sets.powerSet(numbers).stream()
+                    .filter(subSet -> subSet.contains(0))
+                    .filter(subSet -> subSet.size() < numbers.size());
+        }
+
+        JoinEnumerationResult createJoinAccordingToPartitioning(List<PlanNode> sources, List<Symbol> outputSymbols, Set<Integer> partitioning)
+        {
+            Set<PlanNode> leftSources = partitioning.stream()
+                    .map(sources::get)
+                    .collect(toImmutableSet());
+            Set<PlanNode> rightSources = Sets.difference(ImmutableSet.copyOf(sources), ImmutableSet.copyOf(leftSources));
```

Comment:
`Sets.difference` is just a view. Since `rightSources` is used more than once it make sense to materialize it. You can materialized it with a `new HashSet()` as well. `ImmutableSet.copyOf()` seems to be more usual.